jobs/taskrunner_test.go

```diff
@@ -0,0 +1,342 @@
+package jobs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	"golang.org/x/crypto/bcrypt"
+)
+
+func TestNewTaskRunner(t *testing.T) {
+	type args struct {
+		config map[string]interface{}
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *TaskRunner
+		wantErr bool
+	}{
+		{
+			name:    "nil config",
+			wantErr: true,
+		},
+		{
+			name: "empty config",
+			args: args{
+				config: map[string]interface{}{},
+			},
+			wantErr: true,
+		},
+		{
+			name: "endpoint wrong type",
+			args: args{
+				config: map[string]interface{}{
+					"endpoint": 123,
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "endpoint_template wrong type",
+			args: args{
+				config: map[string]interface{}{
+					"endpoint_template": 123,
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "encrypt is wrong type",
+			args: args{
+				config: map[string]interface{}{
+					"endpoint": "http://127.0.0.1:8080/v1/ecs/acct1/cluster/clu1/taskdefs/td1/tasks",
+					"encrypt":  12345,
+				},
+			},
+			want: &TaskRunner{
+				AuthHeader: "X-Auth-Token",
+				Encrypt:    true,
+				Endpoint:   "http://127.0.0.1:8080/v1/ecs/acct1/cluster/clu1/taskdefs/td1/tasks",
+			},
+		},
+		{
+			name: "endpoint and endpoint_template wrong type",
+			args: args{
+				config: map[string]interface{}{
+					"endpoint":         "http://127.0.0.1:8080/v1/ecs/acct1/cluster/clu1/taskdefs/td1/tasks",
+					"endpointTemplate": "http://127.0.0.1:8080/v1/ecs/{{.Account}}/cluster/{{.Cluster}}/taskdefs/{{ .Name }}/tasks",
+				},
+			},
+			want: &TaskRunner{
+				AuthHeader:       "X-Auth-Token",
+				Encrypt:          true,
+				Endpoint:         "http://127.0.0.1:8080/v1/ecs/acct1/cluster/clu1/taskdefs/td1/tasks",
+				EndpointTemplate: "http://127.0.0.1:8080/v1/ecs/{{.Account}}/cluster/{{.Cluster}}/taskdefs/{{ .Name }}/tasks",
```

Comment:
It looks like if both of these are set we should only use `Endpoint`, but we are still setting both?

Reply:
derp - bad test name.  the endpoint and endpointTemplate will both get set, its up to the `Run()` to decide how it gets used.  that's partly why I think we need the `Validate()`. Even if different runners have different requirements, they should at least make sure what's set up will work.